scripts/migrate-portals.js

```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env node
+const fs = require('fs');
+const readline = require('readline');
+const { MatrixClient, LogService } = require('matrix-bot-sdk');
+
+LogService.setLevel('ERROR');
+
+const client = new MatrixClient(process.env.MATRIX_URL, process.env.MATRIX_AT);
+
+const dryRun = process.env.DRY_RUN ? process.env.DRY_RUN === 'true' : true;
+
+const rl = readline.createInterface({
+    input: fs.createReadStream(process.env.ROOM_STORE || './room-store.db'),
+    output: process.stdout,
+    terminal: false
+});
+
+let portalRooms = [];
+
+rl.on('line', (line) => {
+    const roomEntry = JSON.parse(line);
+    if (!roomEntry.matrix_id) {
+        return;
+    }
+    if (!roomEntry || !roomEntry.data || !roomEntry.data.portal) {
+        return;
+    }
+    const gitterRoomId = roomEntry.remote_id;
+    const oldMatrixRoomId = roomEntry.matrix_id;
+    const localpart = gitterRoomId.replace('/', '_');
+    const alias = `#${localpart}:gitter.im`;
+    portalRooms.push({gitterRoomId, oldMatrixRoomId, alias});
+});
+
+async function getNewRoomId(alias) {
+    try {
+        const targetRoomId = await client.resolveRoom(alias);
+        return { targetRoomId, joined: false };
+    } catch (ex) {
+        if (dryRun) {
+            return {
+                targetRoomId: '!noroomyet:gitter.im',
+                joined: false,
+            };
+        }
+        console.log("Joining new gitter room...");
+        await new Promise((r) => setTimeout(r, 5000)); // Joins cost more in rate limits
+        return {
+            targetRoomId: await client.joinRoom(alias),
+            joined: true,
+        };
+    }
+}
+
+rl.on('close', async () => {
+    let index = 0;
+    try {
+        index = parseInt(await fs.promises.readFile('checkpoint.txt', 'utf-8'), 10);
+    } catch (ex) {
+        console.log(`Failed to fetch index, starting from the beginning`);
+    }
+    console.log(`Starting from index ${index}`);
+    for (const {gitterRoomId, oldMatrixRoomId, alias} of portalRooms.slice(index)) {
+        await new Promise((r) => setTimeout(r, 3000));
+        const {targetRoomId, joined} = await getNewRoomId(alias);
+        console.log(`${gitterRoomId} -> ${targetRoomId} (from: ${oldMatrixRoomId})`);
+        if (!targetRoomId) {
+            console.log(`No target room for ${gitterRoomId}!`);
+        }
+        if (!dryRun) {
+            console.log("Joining new room...");
+            await client.joinRoom(alias)
+            console.log("Add tombstone...");
+            await client.sendStateEvent(oldMatrixRoomId, 'm.room.tombstone', '', {
+                body: `The matrix.org Gitter bridge has been discontinued. You can view this channel on the new bridge over at ${alias}`,
+                replacement_room: targetRoomId,
+            });
```

Comment:
The only thing I don't see here from the [plan](https://gitlab.com/gitterHQ/webapp/-/issues/2675#note_462793676) is to remove the room alias. Do we still want to do that here?

Reply:
good point, let's do it